db.go

```diff
@@ -477,190 +502,177 @@ func (db *DB) recover(ctx context.Context) error {
-				record, err := reader.Read()
-				if err != nil {
-					return fmt.Errorf("read record: %w", err)
+			case *walpb.Entry_Write_:
+				entry := e.Write
+				tableName := entry.TableName
+				if lastPersistedTx, ok := persistedTables[tableName]; ok && tx < lastPersistedTx {
+					// This write has already been successfully persisted, so we can
+					// skip it.
+					return nil
 				}
 
-				if err := table.active.InsertRecord(ctx, tx, record); err != nil {
-					return fmt.Errorf("insert record into block: %w", err)
+				table, err := db.GetTable(tableName)
+				var tableErr ErrTableNotFound
+				if errors.As(err, &tableErr) {
+					// This means the WAL was truncated at a point where this write
+					// was already successfully persisted to disk in more optimized
+					// form than the WAL.
+					return nil
 				}
-			default:
-				serBuf, err := dynparquet.ReaderFromBytes(entry.Data)
 				if err != nil {
-					return fmt.Errorf("deserialize buffer: %w", err)
+					return fmt.Errorf("get table: %w", err)
 				}
 
-				if err := table.active.Insert(ctx, tx, serBuf); err != nil {
-					return fmt.Errorf("insert buffer into block: %w", err)
+				switch e.Write.Arrow {
+				case true:
+					reader, err := ipc.NewReader(bytes.NewReader(entry.Data))
+					if err != nil {
+						return fmt.Errorf("create ipc reader: %w", err)
+					}
+					record, err := reader.Read()
+					if err != nil {
+						return fmt.Errorf("read record: %w", err)
+					}
+
+					if err := table.active.InsertRecord(ctx, tx, record); err != nil {
+						return fmt.Errorf("insert record into block: %w", err)
+					}
+				default:
+					serBuf, err := dynparquet.ReaderFromBytes(entry.Data)
+					if err != nil {
+						return fmt.Errorf("deserialize buffer: %w", err)
+					}
+
+					if err := table.active.Insert(ctx, tx, serBuf); err != nil {
+						return fmt.Errorf("insert buffer into block: %w", err)
+					}
 				}
-			}
 
-			// After every insert we're setting the tx and highWatermark to the replayed tx.
-			// This allows the block's compaction to start working on the inserted data.
-			db.tx.Store(tx)
-			db.highWatermark.Store(tx)
-		case *walpb.Entry_TableBlockPersisted_:
-			return nil
-		case *walpb.Entry_Snapshot_:
+				// After every insert we're setting the tx and highWatermark to the replayed tx.
+				// This allows the block's compaction to start working on the inserted data.
+				db.tx.Store(tx)
+				db.highWatermark.Store(tx)
+			case *walpb.Entry_TableBlockPersisted_:
+				return nil
+			case *walpb.Entry_Snapshot_:
+				return nil
+			default:
+				return fmt.Errorf("unexpected WAL entry type: %t", e)
+			}
 			return nil
 		default:
-			return fmt.Errorf("unexpected WAL entry type: %t", e)
+			return fmt.Errorf("unexpected number of replay counts; expected 2, got %v", i)
 		}
-		return nil
-	}); err != nil {
-		return fmt.Errorf("second WAL replay: %w", err)
 	}
-
-	level.Info(db.logger).Log("msg", "replaying WAL completed", "duration", time.Since(start))
```

Comment:
These log messages have been pretty useful (I think they're not in the latest version of this PR)

Reply:
Yes! Thank you, I meant to include that again but forgot I had deleted it

Reply:
added back in